src/kernel.ts

```diff
@@ -1,1069 +1,802 @@
-        var msg_type = header.msg_type;
-        if (msg_type !== 'input_request') {
-            console.log("Invalid input request!", request);
-            return;
-        }
-        var callbacks = this.get_callbacks_for_msg(request.parent_header.msg_id);
-        if (callbacks) {
-            if (callbacks.input) {
-                callbacks.input(request);
-            }
-        }
-    }
+/**
+ * Bit flags for the Kernel future state.
+ */
+enum KernelFutureFlag {
+  GotReply = 0x1,
+  GotIdle = 0x2,
+  AutoDispose = 0x4,
+  IsDone = 0x8
+}
 
-    public id: string;
-    public name: string;
-    public kernel_service_url: string;
-    public kernel_url: string;
-    public ws_url: string;
-    public username: string;
-    public session_id: string;
-    public ws: WebSocket;
-    public info_reply: any;
-    public WebSocket: any;
-    public comm_manager: comm.CommManager;
-    public last_msg_id: string;
-    public last_msg_callbacks: any;
-    public reconnect_limit: number;
-
-    private _msg_callbacks: any;
-    private _msg_queue: Promise;
-    private _autorestart_attempt: number;
-    private _reconnect_attempt: number;
-    private _iopub_handlers: any;
 
+/**
+ * Implementation of a Kernel Future.
+ */
+class KernelFutureHandler extends Disposable implements IKernelFuture {
+
+  /**
+   * Dispose and unregister the future.
+   */
+  dispose(): void {
+    super.dispose();
+    this._input = null;
+    this._output = null;
+    this._reply = null;
+    this._done = null;
+  }
+
+  /**
+   * Register a reply handler. Returns `this`.
+   */
+  onReply(cb: (msg: IKernelMsg) => void): IKernelFuture {
+    this._reply = cb;
+    return this;
+  }
+
+  /**
+   * Register an output handler. Returns `this`.
+   */
+  onOutput(cb: (msg: IKernelMsg) => void): IKernelFuture {
+    this._output = cb;
+    return this;
+  }
+
+  /**
+   * Register a done handler. Returns `this`.
+   */
+  onDone(cb: (msg: IKernelMsg) => void): IKernelFuture {
+    this._done = cb;
+    return this;
+  }
+
+  /**
+   * Register an input handler. Returns `this`.
+   */
+  onInput(cb: (msg: IKernelMsg) => void): IKernelFuture {
+    this._input = cb;
+    return this;
+  }
+
+  /**
+   * Get the current autoDispose status of the future..
+   */
+  get autoDispose(): boolean {
+    return this._testFlag(KernelFutureFlag.AutoDispose);
+  }
+
+  /**
+   * Set the current autoDispose behavior of the future.
+   *
+   * If True, it will self-dispose() after onDone() is called.
+   */
+  set autoDispose(value: boolean) {
+    if (value) {
+      this._setFlag(KernelFutureFlag.AutoDispose);
+    } else {
+      this._clearFlag(KernelFutureFlag.AutoDispose);
+    }
+  }
+
+  /**
+   * Check for message done state.
+   */
+  get isDone(): boolean {
+    return this._testFlag(KernelFutureFlag.IsDone);
+  }
+
+  /**
+   * Handle an incoming message from the kernel belonging to this future.
+   */
+  handleMsg(msg: IKernelMsg): void {
```

Comment:
We should think about what we want to do if a message is received after 'done'. If the future isn't auto-disposed, should it continue to process messages? And if so, should 'done' be called multiple times? i.e. do we clear the reply and done handlers after the first 'done', leaving only 'output'?